src/lenskit/reranking/fair.py

```diff
@@ -0,0 +1,223 @@
+                    self.pmf_cache[block_size] = binom.pmf(np.arange(block_size + 1), block_size, p)
+
+                S = np.convolve(self.pmf_cache[block_size], S)
+                # set the next index (j-1) to zero since the last element of each block
+                # increases by 1 and now we need at least j protected up to this point
+                S[j - 1] = 0
+
+        return float(1 - S.sum())
+
+    def binary_search_significance(self, k, p, alpha, tolerance=1e-10, max_iter=100):
+        low, high = 0, 1
+
+        for _ in range(max_iter):
+            alpha_c = (low + high) / 2
+            rejection_prob = self._compute_rejection_prob(k, p, alpha_c)
+
+            if rejection_prob > alpha:
+                high = alpha_c
+            else:
+                low = alpha_c
+            if abs(rejection_prob - alpha) < tolerance or high - low < tolerance:
+                break
+
+        alpha_c = (low + high) / 2
+        return alpha_c
+
+    @override
+    def train(self, data: Dataset, options: TrainingOptions = TrainingOptions()):
+        """
+        Precompute adjusted alpha and m-table for the configured (k,p,alpha)
+
+        Training computes the (multiple-tests–corrected) prefix requirements m[i]
+        so that every prefix of size i has at least m[i] protected items with
+        significance α (after correction).
+
+        Args:
+            protected attributes:
+                The item entity table must have a scalar attribute ``protected`` that is ``True``
+                if the item is in the protected group and ``False`` otherwise.
+        """
+        self.pmf_cache = {}
+
+        log = _log.bind(alpha=self.config.alpha, k_max=self.config.k)
+        timer = Stopwatch()
+        log.info("begining FA*IR Tuning")
+
+        # 1. find α_c
+        self.alpha_c_ = self.binary_search_significance(
+            k=self.config.k, p=self.config.p, alpha=self.config.alpha
+        )
+        # 2. compute m_table
+        self.m_list_ = self._compute_m_list(k=self.config.k, p=self.config.p, alpha=self.alpha_c_)
+
+        items = data.entities("item")
+
+        if "protected" not in items.attributes:
+            raise ValueError("Dataset is missing required 'protected' attribute for item entities")
+
+        self.protected_attributes = items.attribute("protected").pandas()
+
+        log.info(
+            "[%s] tuned with p=%f, alpha=%f → alpha_c=%.8f",
+            timer,
+            self.config.p,
+            self.config.alpha,
+            self.alpha_c_,
+        )
+
+    def __call__(
+        self,
+        items: ItemList,
+    ) -> ItemList:
+        item_ids = items.ids()
+
+        p_items = deque()
+        up_items = deque()
+
+        is_protected = self.protected_attributes.reindex(item_ids).fillna(False).values
+
+        p_items = deque(np.nonzero(is_protected)[0])
+        up_items = deque(np.nonzero(~is_protected)[0])
+
+        count_prot = 0
+        reranked_indices = deque()
```

Comment:
since we're only appending, not popping from the front, we can just use a list here instead of a `deque`.